examples/DivisorCurve.jl

```diff
@@ -0,0 +1,327 @@
+export AffineCurveDivisor, ProjCurveDivisor, iseffective, multiplicity,
+       divisor
+
+################################################################################
+
+abstract type CurveDivisor end
+
+################################################################################
+# The curve C is assume to be smooth and irreducible, but it is not checked.
+@doc Markdown.doc"""
+    AffineCurveDivisor(C::AffinePlaneCurve{S}, D::Dict{Point{S}, Int}) where S <: FieldElem
+
+Given a curve `C` which is assumed to be smooth and irreducible, return the divisor on the curve `C` defined by `D`.
+"""
+mutable struct AffineCurveDivisor{S <: FieldElem} <: CurveDivisor
+    C::AffinePlaneCurve{S}
+    divisor::Dict{Point{S}, Int}
+    degree::Int
+    function AffineCurveDivisor(C::AffinePlaneCurve{S}, D::Dict{Point{S}, Int}) where S <: FieldElem
+        for P in keys(D)
+            iszero(evaluate(C.eq, P.coord)) || error("The point ", P.coord, "is not on the curve.")
+            if iszero(D[P])
+                delete!(D, P)
+            end
+        end
+        new{S}(C,
+            D,
+            sum(values(D)))
+    end
+end
+
+function AffineCurveDivisor(C::AffinePlaneCurve{S}, P::Point{S}, m::Int=1) where S <: FieldElem
+    iszero(evaluate(C.eq, P.coord)) || error("The point is not on the curve.")
+    return AffineCurveDivisor(C, Dict(P => m))
+end
+
+################################################################################
+
+@doc Markdown.doc"""
+    ProjCurveDivisor(C::ProjPlaneCurve{S}, D::Dict{Oscar.Geometry.ProjSpcElem{S}, Int}) where S <: FieldElem
+
+Given a curve `C` which is assumed to be smooth and irreducible, return the divisor on the curve `C` defined by `D`.
+"""
+mutable struct ProjCurveDivisor{S <: FieldElem} <: CurveDivisor
+    C::ProjPlaneCurve{S}
+    divisor::Dict{Oscar.Geometry.ProjSpcElem{S}, Int}
+    degree::Int
+    function ProjCurveDivisor(C::ProjPlaneCurve{S}, D::Dict{Oscar.Geometry.ProjSpcElem{S}, Int}) where S <: FieldElem
+        for P in keys(D)
+            iszero(evaluate(C.eq, P.v)) || error("The point ", P, "is not on the curve.")
+            if iszero(D[P])
+                delete!(D, P)
+            end
+        end
+        new{S}(C,
+            D,
+            sum(values(D)))
+    end
+end
+#function Base.show(io::IO, D::ProjCurveDivisor)
+#    println(io, "On the curve ", D.C.eq, ", divisor defined by ", D.divisor)
+#end
+
+function ProjCurveDivisor(C::ProjPlaneCurve{S}, P::Oscar.Geometry.ProjSpcElem{S}, m::Int=1) where S <: FieldElem
+    iszero(evaluate(C.eq, P.v)) || error("The point is not on the curve.")
+    return ProjCurveDivisor(C, Dict(P => m))
+end
+
+################################################################################
+# To get a nice output for the divisors.
+
+function AbstractAlgebra.needs_parentheses(A::Array{T, 1}) where T <: RingElem
```

Comment:
Do you remember why this was necessary?

Reply:
Without it there was an error...

Reply:
As there is no test here, what command was exactly failing?

Reply:
Without the function needs_parentheses, when I try to create a AffineCurveDivisor, I get the error : Error showing value of type AffineCurveDivisor{fmpq}:
ERROR: MethodError: no method matching needs_parentheses(::Array{fmpq,1}).
It seems to come from expressify, which apparently wants to know if it needs brackets or not.

Reply:
Thanks, this information helps.